lib/api/src/module.rs

```diff
@@ -251,6 +251,9 @@ impl Module {
         Ok(Self::from_artifact(store, artifact))
     }
 
+    /// Low-level constructor. It must be used to build a unique
+    /// `Module`. It must not be used to build back a `Module` from an
+    /// already compiled `Artifact`.
```

Comment:
Depending on how bad the consequences for misuse are here, we should consider marking this as `unsafe`.